sky/provision/vast/utils.py

```diff
@@ -0,0 +1,103 @@
+# pylint: disable=assignment-from-no-return
+"""Vast library wrapper for SkyPilot."""
+from typing import Any, Dict, List, Optional
+
+from sky import sky_logging
+from sky.adaptors import vast
+
+logger = sky_logging.init_logger(__name__)
+
+
+def list_instances() -> Dict[str, Dict[str, Any]]:
+    """Lists instances associated with API key."""
+    instances = vast.vast().show_instances()
+
+    instance_dict: Dict[str, Dict[str, Any]] = {}
+    for instance in instances:
+        instance['id'] = str(instance['id'])
+        info = instance.copy()
+
+        if isinstance(instance['actual_status'], str):
+            info['status'] = instance['actual_status'].upper()
+        else:
+            info['status'] = 'UNKNOWN'
+        info['name'] = instance['label']
+
+        instance_dict[instance['id']] = info
+
+    return instance_dict
+
+
+def launch(name: str, instance_type: str, region: str, disk_size: int,
+           image_name: str, ports: Optional[List[int]], public_key: str) -> str:
+    """Launches an instance with the given parameters.
+
+    Converts the instance_type to the Vast GPU name, finds the specs for the
+    GPU, and launches the instance.
+    """
+    del ports
+    del public_key
+
+    gpu_name = instance_type.split('-')[1].replace('_', ' ')
+    num_gpus = int(instance_type.split('-')[0].replace('x', ''))
```

Comment:
Is it possible to have a comment for the format of vast ai's instance name format?

Also, it will be great to use regex instead of `split` to enhance the robustness of our code ;)

Reply:
We don't have "instance name format". We have uniquely assigned snowflake machines that are listable through an expressive query syntax so this is an invention for skypilot. As I said elsewhere, I'm thinking about pushing for at least a hybrid approach of an optional categorization to match with how most other providers slice their inventory.

Reply:
I brought up the idea of creating an abstract instance type like other providers have. I just floated it, I need to get a buy in from the dev team to actually surface it. It's a reasonable idea and I'd be happy to revisit this block of code if/when it happens

Reply:
Oh I mean, it looks like the "instance_type"'s name always be the format of `2x-A100` (so you are splitting using `-` and then parse them separately). Is that true? If it is true, can we add a comment for reference link (in your documentation)? If it is not true, what will happens if the name not following the format? will it just throw an error here? should we have error handling for this?

Reply:
Since our "instance type" concept was invented to conform to the catalog format, I can guarantee that. We will eventually be introducing more proper instance types and when that happens, if we need to change things, I'll happily take responsibility for the transition.

Reply:
Got it. If that is the case, maybe worth referencing the catalog code that generate this "instance type" string :))

Reply:
fine. https://github.com/skypilot-org/skypilot/pull/4365/commits/f2d838134ff8668ac50217aac56aba3ed257a61e